reqlog/external_latency.go

```diff
@@ -0,0 +1,102 @@
+// Copyright © 2024 Ory Corp
+// SPDX-License-Identifier: Apache-2.0
+
+package reqlog
+
+import (
+	"context"
+	"net/http"
+	"sync"
+	"time"
+)
+
+// WithEnableExternalLatencyMeasurement returns a context that measures external latencies.
+func WithEnableExternalLatencyMeasurement(ctx context.Context) context.Context {
+	container := contextContainer{
+		latencies: make([]externalLatency, 0),
+	}
+	return context.WithValue(ctx, internalLatencyKey, &container)
+}
+
+// ExternalCallsMiddleware is a middleware that sets up the request context to measure external calls.
+// It has to be used before any other middleware that reads the final external latency.
+func ExternalCallsMiddleware(rw http.ResponseWriter, r *http.Request, next http.HandlerFunc) {
```

Comment:
This does not need an extra middleware, we can just add the context key in the main reqlog middleware.